src/test/scala/cromwell/binding/values/WdlValueSpec.scala

```diff
@@ -0,0 +1,101 @@
+    (WdlFloat(Double.NegativeInfinity), "-Infinity"),
+    (WdlInteger(0), "0"),
+    (WdlInteger(Int.MaxValue), "2147483647"),
+    (WdlInteger(Int.MinValue), "-2147483648"),
+    (WdlString(""), ""),
+    (WdlString("test\\'/string"), "test\\'/string"))
+
+  forAll(wdlValueRawStrings) { (wdlValue, rawString) =>
+    it should s"exactly convert a Wdl${wdlValue.wdlType.toWdlString} to/from raw string '$rawString'" in {
+      val toRawString = wdlValue.toRawString
+      toRawString should be(rawString)
+
+      val wdlType = wdlValue.wdlType
+      val fromRawString = wdlType.fromRawString(toRawString)
+      fromRawString should be(wdlValue)
+      fromRawString.wdlType should be(wdlType)
+    }
+  }
+
+  val wdlFloatTruncated = Table(
+    ("wdlValue", "rawString", "validateFloat"),
+    (WdlFloat(Double.MinPositiveValue), "4.9E-324", { d: Double => d == 0.0 }),
+    (WdlFloat(Double.NaN), "NaN", { d: Double => d.isNaN }),
+    (WdlFloat(Double.MaxValue), "1.7976931348623157E308", { d: Double => d.isPosInfinity }),
+    (WdlFloat(Double.MinValue), "-1.7976931348623157E308", { d: Double => d.isNegInfinity }))
+
+  forAll(wdlFloatTruncated) { (wdlValue, rawString, validateFloat) =>
+    it should s"truncate a Wdl${wdlValue.wdlType.toWdlString} to/from raw string '$rawString'" in {
+      val toRawString = wdlValue.toRawString
+      toRawString should be(rawString)
+
+      val wdlType = wdlValue.wdlType
+      val fromRawString = wdlType.fromRawString(toRawString)
+      fromRawString shouldNot be(wdlValue)
+      validateFloat(fromRawString.value) should be(right = true)
+      fromRawString.wdlType should be(wdlType)
+    }
+  }
+
+  val wdlExpressionRawStrings = Table(
+    ("wdlValue", "rawString"),
+    (WdlExpression.fromString(" 1 != 0 "), "1 != 0"),
+    (WdlExpression.fromString("10 % 3.5"), "10 % 3.5"),
+    (WdlExpression.fromString("10 % 3"), "10 % 3"),
+    (WdlExpression.fromString("10-6.7"), "10 - 6.7"),
+    (WdlExpression.fromString(""" 1 + "String" """), """1 + "String""""),
+    (WdlExpression.fromString("a + b"), "a + b"),
+    (WdlExpression.fromString("a(b, c)"), "a(b, c)"),
+    (WdlExpression.fromString("\"a\" + \"b\""), "\"a\" + \"b\""),
+    (WdlExpression.fromString("a.b.c"), "a.b.c"))
+
+  forAll(wdlExpressionRawStrings) { (wdlValue, rawString) =>
+    it should s"resemble a Wdl${wdlValue.wdlType.toWdlString} to/from raw string '$rawString'" in {
+      val toRawString = wdlValue.toRawString
+      toRawString should be(rawString)
+
+      val wdlType = wdlValue.wdlType
+      val fromRawString = wdlType.fromRawString(toRawString)
+      fromRawString shouldNot be(wdlValue)
+      fromRawString.toString should be(wdlValue.toString)
+      fromRawString.wdlType should be(wdlType)
+    }
+  }
+
+  val notImplementRawString = Table(
+    "wdlValue",
+    WdlObject(Map("key" -> WdlString("value"))),
+    WdlNamespace.load(SampleWdl.HelloWorld.wdlSource()))
+
+  forAll(notImplementRawString) { wdlValue =>
```

Comment:
nm, on further reflection I like this.  :smile:  If these types ever gain real raw string implementations this will force the implementer to address the testing issue.